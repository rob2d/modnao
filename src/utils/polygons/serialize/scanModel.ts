import S from '@/constants/StructSizes';
import {
  nlColoredVertexConversions,
  nlMeshConversions,
  nlModelConversions,
  nlPolygonConversions,
  nlVertexConversions
} from './NLPropConversionDefs';
import { processNLConversions } from './processNLConversions';

export type ScanModelParams = {
  address: number;
  ramAddress: number;
  buffer: SharedArrayBuffer;
  index: number;
};

export default function scanModel({
  address,
  ramAddress,
  buffer,
  index
}: ScanModelParams) {
  const wBuffer = Buffer.from(new Uint8Array(buffer));
  // (1) scan base model props
  const model = processNLConversions<NLModel>(
    nlModelConversions,
    wBuffer,
    address
  );

  model.address = address;
  model.ramAddress = ramAddress;
  model.meshes = [];

  // total vert count is populated
  // at the end of parsing polygons
  // following a ds flag

  model.totalVertexCount = 0;

  // (2) scan through meshes
  try {
    let detectedModelEnd = false;
    let structAddress = address + S.MODEL_HEADER;

    while (structAddress < wBuffer.length && !detectedModelEnd) {
      if (wBuffer.readUInt32LE(structAddress) === 0) {
        detectedModelEnd = true;
        structAddress += 4;
        model.totalVertexCount = wBuffer.readUInt32LE(structAddress);
        structAddress += 4;
        break;
      }

      const mesh = processNLConversions<NLMesh>(
        nlMeshConversions,
        wBuffer,
        structAddress
      );

      mesh.polygons = [];

      structAddress += S.MESH;
      const meshEndAddress = structAddress + mesh.polygonDataLength;

      // (3) scan polygons within mesh
      while (
        structAddress < meshEndAddress &&
        structAddress + S.VERTEX_B < wBuffer.length &&
        !detectedModelEnd
      ) {
        const polygon = processNLConversions<NLPolygon>(
          nlPolygonConversions,
          wBuffer,
          structAddress
        );
        polygon.vertices = [];

        structAddress = polygon.address + S.POLYGON_HEADER;

        // (4) scan vertices within polygon
        let detectedMeshEnd = false;
        for (let i = 0; i < polygon.actualVertexCount; i++) {
          // detect modelEnd flag in the vertex position
          if (wBuffer.readUInt32LE(structAddress) === 0) {
            detectedModelEnd = true;
            detectedMeshEnd = true;
            structAddress += 4;
            model.totalVertexCount = wBuffer.readUInt32LE(structAddress);
            structAddress += 4;
            break;
          }

          if (detectedMeshEnd) {
            break;
          }

          if (structAddress + S.VERTEX_B >= wBuffer.length) {
            console.log(
              `invalid logic in parser occurred somewhere near ${structAddress}; went beyond buffer length of ${wBuffer.length}`
            );
            break;
          }

          const vertex = processNLConversions<NLVertex>(
            mesh.hasColoredVertices
              ? nlColoredVertexConversions
              : nlVertexConversions,
            wBuffer,
            structAddress
          );

          vertex.index = i;
          polygon.vertices.push(vertex);
          structAddress +=
            vertex.addressingMode === 'direct' ? S.VERTEX_A : S.VERTEX_B;

          if (structAddress >= meshEndAddress) {
            detectedMeshEnd = true;
          }
        }

        // populate polygon indices
        // note: there are some quirks that aren't fully sorted
        // here with winding order. Good case scenario to debug
        // is bird model on STG01 in MVC2
        let stripCount = 0;
        const indices: number[] = [];
        polygon.vertices.forEach((v, i) => {
          if (polygon.vertexGroupMode === 'regular') {
            if (i > polygon.vertices.length - 3) {
              return;
            }

            if (stripCount % 2 === 0) {
              if (polygon.flags.cullingType === 'front') {
                indices.push(i + 1, i, i + 2);
              } else {
                indices.push(i, i + 1, i + 2);
              }
            } else {
              if (polygon.flags.cullingType === 'front') {
                indices.push(i, i + 1, i + 2);
              } else {
                indices.push(i + 1, i, i + 2);
              }
            }
          }

          stripCount += 1;
        });

        if (polygon.vertexGroupMode === 'triple') {
          for (let i = 2; i < polygon.vertices.length; i += 3) {
            if (polygon.flags.cullingType === 'front') {
              indices.push(i - 1, i - 2, i);
            } else {
              indices.push(i - 2, i - 1, i);
            }
          }
          stripCount += 1;
        }

        // pre-calculate order of triangle indices
        // for wireframe drawing & uv atlas polygon highlights
        const triIndices: number[] = [];

        for (let i = 0; i < indices.length; i += 3) {
          triIndices.push(
            indices[i],
            indices[i + 1],
            indices[i + 2],
            indices[i]
          );
        }

        polygon.indices = indices;
        polygon.triIndices = triIndices;
        mesh.polygons.push(polygon);
      }

      model.meshes.push(mesh);
    }
  } catch (error) {
    // @TODO: more thoughtful handling of errors
    console.trace(
      `${index}: error parsing model @ 0x${address.toString(16)}`,
      error
    );
  }

  return model;
}
