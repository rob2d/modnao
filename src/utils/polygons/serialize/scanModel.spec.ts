import { promises as fs } from 'fs';
import path from 'path';

import scanModel from './scanModel';

// @TODO use model generator for mocking when this exists

describe('scanModel', () => {
  it('extracts a test stage file predictably', async () => {
    const fileBuffer = await fs.readFile(
      path.join(process.cwd(), 'src/__mocks__/STGXXPOL.BIN')
    );

    const buffer = new SharedArrayBuffer(fileBuffer.byteLength);
    new Uint8Array(buffer).set(new Uint8Array(fileBuffer));
    const modelData = scanModel({
      buffer,
      address: 528,
      index: 0,
      ramAddress: 0
    });

    expect(modelData).toMatchSnapshot('scan-model-model-data');
  });
});
