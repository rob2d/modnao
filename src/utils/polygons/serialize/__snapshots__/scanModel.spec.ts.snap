// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scanModel extracts a test stage file predictably: scan-model-model-data 1`] = `
{
  "address": 528,
  "meshes": [
    {
      "address": 552,
      "alpha": 1,
      "baseParams": 2147483692,
      "color": [
        1,
        1,
        1,
      ],
      "hasColoredVertices": false,
      "isOpaque": true,
      "polygonDataLength": 136,
      "polygons": [
        {
          "actualVertexCount": 4,
          "address": 632,
          "flags": {
            "culling": true,
            "cullingType": "front",
            "envMaps": false,
            "gouradShading": false,
            "reuseGlobalParams": false,
            "spriteQuad": true,
            "strip": true,
            "superVertexIndex": false,
            "triangles": false,
          },
          "indices": [
            1,
            0,
            2,
            1,
            2,
            3,
          ],
          "triIndices": [
            1,
            0,
            2,
            1,
            1,
            2,
            3,
            1,
          ],
          "vertexCount": 4,
          "vertexGroupMode": "regular",
          "vertexGroupModeValue": 22,
          "vertices": [
            {
              "address": 640,
              "addressingMode": "direct",
              "contentAddress": 640,
              "contentModeValue": 3304718337,
              "index": 0,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                -2000.0001220703125,
                1400,
                -2000,
              ],
              "uv": [
                0,
                -1.401298464324817e-45,
              ],
            },
            {
              "address": 672,
              "addressingMode": "direct",
              "contentAddress": 672,
              "contentModeValue": 3304718337,
              "index": 1,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                -2000.0001220703125,
                1400,
                499.99969482421875,
              ],
              "uv": [
                10,
                -1.401298464324817e-45,
              ],
            },
            {
              "address": 704,
              "addressingMode": "direct",
              "contentAddress": 704,
              "contentModeValue": 3304718337,
              "index": 2,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                -2000.0001220703125,
                0,
                -2000,
              ],
              "uv": [
                0,
                -10.000000953674316,
              ],
            },
            {
              "address": 736,
              "addressingMode": "direct",
              "contentAddress": 736,
              "contentModeValue": 3304718337,
              "index": 3,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                -2000.0001220703125,
                0,
                499.99969482421875,
              ],
              "uv": [
                10,
                -10.000000953674316,
              ],
            },
          ],
        },
      ],
      "position": [
        0,
        0,
        -1000.0003051757812,
      ],
      "specularAlpha": 0,
      "specularColor": [
        1,
        1,
        1,
      ],
      "textureColorFormat": "RGB555",
      "textureColorFormatValue": 8,
      "textureControlValue": 134217728,
      "textureIndex": 0,
      "textureInstructions": 2197815296,
      "textureSize": [
        64,
        64,
      ],
      "textureSizeValue": 91,
      "textureWrappingFlags": {
        "hFlip": false,
        "hRepeat": true,
        "hStretch": false,
        "vFlip": false,
        "vRepeat": true,
      },
      "textureWrappingValue": 8,
      "vertexColorModeValue": -1,
    },
    {
      "address": 768,
      "alpha": 1,
      "baseParams": 2147483708,
      "color": [
        1,
        1,
        1,
      ],
      "hasColoredVertices": false,
      "isOpaque": true,
      "polygonDataLength": 136,
      "polygons": [
        {
          "actualVertexCount": 4,
          "address": 848,
          "flags": {
            "culling": true,
            "cullingType": "front",
            "envMaps": false,
            "gouradShading": false,
            "reuseGlobalParams": false,
            "spriteQuad": true,
            "strip": true,
            "superVertexIndex": false,
            "triangles": false,
          },
          "indices": [
            1,
            0,
            2,
            1,
            2,
            3,
          ],
          "triIndices": [
            1,
            0,
            2,
            1,
            1,
            2,
            3,
            1,
          ],
          "vertexCount": 4,
          "vertexGroupMode": "regular",
          "vertexGroupModeValue": 22,
          "vertices": [
            {
              "address": 856,
              "addressingMode": "direct",
              "contentAddress": 856,
              "contentModeValue": 3292233729,
              "index": 0,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                -750.0000610351562,
                0,
                -750,
              ],
              "uv": [
                0,
                -1.401298464324817e-45,
              ],
            },
            {
              "address": 888,
              "addressingMode": "direct",
              "contentAddress": 888,
              "contentModeValue": 3292233729,
              "index": 1,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                -750.0000610351562,
                250,
                -750,
              ],
              "uv": [
                10,
                -1.401298464324817e-45,
              ],
            },
            {
              "address": 920,
              "addressingMode": "direct",
              "contentAddress": 920,
              "contentModeValue": 1144750081,
              "index": 2,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                750.0000610351562,
                0,
                -750,
              ],
              "uv": [
                0,
                -10.000000953674316,
              ],
            },
            {
              "address": 952,
              "addressingMode": "direct",
              "contentAddress": 952,
              "contentModeValue": 1144750081,
              "index": 3,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                750.0000610351562,
                250,
                -750,
              ],
              "uv": [
                10,
                -10.000000953674316,
              ],
            },
          ],
        },
      ],
      "position": [
        0,
        0,
        -1000.0003051757812,
      ],
      "specularAlpha": 0,
      "specularColor": [
        1,
        1,
        1,
      ],
      "textureColorFormat": "RGB555",
      "textureColorFormatValue": 8,
      "textureControlValue": 134217728,
      "textureIndex": 0,
      "textureInstructions": 2197815296,
      "textureSize": [
        64,
        64,
      ],
      "textureSizeValue": 91,
      "textureWrappingFlags": {
        "hFlip": false,
        "hRepeat": true,
        "hStretch": false,
        "vFlip": false,
        "vRepeat": true,
      },
      "textureWrappingValue": 8,
      "vertexColorModeValue": -1,
    },
    {
      "address": 984,
      "alpha": 1,
      "baseParams": 2147483708,
      "color": [
        1,
        1,
        1,
      ],
      "hasColoredVertices": false,
      "isOpaque": true,
      "polygonDataLength": 136,
      "polygons": [
        {
          "actualVertexCount": 4,
          "address": 1064,
          "flags": {
            "culling": true,
            "cullingType": "front",
            "envMaps": false,
            "gouradShading": false,
            "reuseGlobalParams": false,
            "spriteQuad": true,
            "strip": true,
            "superVertexIndex": false,
            "triangles": false,
          },
          "indices": [
            1,
            0,
            2,
            1,
            2,
            3,
          ],
          "triIndices": [
            1,
            0,
            2,
            1,
            1,
            2,
            3,
            1,
          ],
          "vertexCount": 4,
          "vertexGroupMode": "regular",
          "vertexGroupModeValue": 22,
          "vertices": [
            {
              "address": 1072,
              "addressingMode": "direct",
              "contentAddress": 1072,
              "contentModeValue": 1157234689,
              "index": 0,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                2000.0001220703125,
                1400,
                -2000,
              ],
              "uv": [
                0,
                -1.401298464324817e-45,
              ],
            },
            {
              "address": 1104,
              "addressingMode": "direct",
              "contentAddress": 1104,
              "contentModeValue": 1157234689,
              "index": 1,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                2000.0001220703125,
                1400,
                499.99969482421875,
              ],
              "uv": [
                10,
                -1.401298464324817e-45,
              ],
            },
            {
              "address": 1136,
              "addressingMode": "direct",
              "contentAddress": 1136,
              "contentModeValue": 1157234689,
              "index": 2,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                2000.0001220703125,
                0,
                -2000,
              ],
              "uv": [
                0,
                -10.000000953674316,
              ],
            },
            {
              "address": 1168,
              "addressingMode": "direct",
              "contentAddress": 1168,
              "contentModeValue": 1157234689,
              "index": 3,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                2000.0001220703125,
                0,
                499.99969482421875,
              ],
              "uv": [
                10,
                -10.000000953674316,
              ],
            },
          ],
        },
      ],
      "position": [
        0,
        0,
        -1000.0003051757812,
      ],
      "specularAlpha": 0,
      "specularColor": [
        1,
        1,
        1,
      ],
      "textureColorFormat": "RGB555",
      "textureColorFormatValue": 8,
      "textureControlValue": 134217728,
      "textureIndex": 0,
      "textureInstructions": 2197815296,
      "textureSize": [
        64,
        64,
      ],
      "textureSizeValue": 91,
      "textureWrappingFlags": {
        "hFlip": false,
        "hRepeat": true,
        "hStretch": false,
        "vFlip": false,
        "vRepeat": true,
      },
      "textureWrappingValue": 8,
      "vertexColorModeValue": -1,
    },
    {
      "address": 1200,
      "alpha": 1,
      "baseParams": 2147483708,
      "color": [
        1,
        1,
        1,
      ],
      "hasColoredVertices": false,
      "isOpaque": true,
      "polygonDataLength": 136,
      "polygons": [
        {
          "actualVertexCount": 4,
          "address": 1280,
          "flags": {
            "culling": true,
            "cullingType": "front",
            "envMaps": false,
            "gouradShading": false,
            "reuseGlobalParams": false,
            "spriteQuad": true,
            "strip": true,
            "superVertexIndex": false,
            "triangles": false,
          },
          "indices": [
            1,
            0,
            2,
            1,
            2,
            3,
          ],
          "triIndices": [
            1,
            0,
            2,
            1,
            1,
            2,
            3,
            1,
          ],
          "vertexCount": 4,
          "vertexGroupMode": "regular",
          "vertexGroupModeValue": 22,
          "vertices": [
            {
              "address": 1288,
              "addressingMode": "direct",
              "contentAddress": 1288,
              "contentModeValue": 3304718337,
              "index": 0,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                -2000.0001220703125,
                0,
                499.99969482421875,
              ],
              "uv": [
                0,
                -1.401298464324817e-45,
              ],
            },
            {
              "address": 1320,
              "addressingMode": "direct",
              "contentAddress": 1320,
              "contentModeValue": 3304718337,
              "index": 1,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                -2000.0001220703125,
                0,
                -2000,
              ],
              "uv": [
                10,
                -1.401298464324817e-45,
              ],
            },
            {
              "address": 1352,
              "addressingMode": "direct",
              "contentAddress": 1352,
              "contentModeValue": 1157234689,
              "index": 2,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                2000.0001220703125,
                0,
                499.99969482421875,
              ],
              "uv": [
                0,
                -10.000000953674316,
              ],
            },
            {
              "address": 1384,
              "addressingMode": "direct",
              "contentAddress": 1384,
              "contentModeValue": 1157234689,
              "index": 3,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                2000.0001220703125,
                0,
                -2000,
              ],
              "uv": [
                10,
                -10.000000953674316,
              ],
            },
          ],
        },
      ],
      "position": [
        0,
        0,
        -1000.0003051757812,
      ],
      "specularAlpha": 0,
      "specularColor": [
        1,
        1,
        1,
      ],
      "textureColorFormat": "RGB555",
      "textureColorFormatValue": 8,
      "textureControlValue": 134217728,
      "textureIndex": 0,
      "textureInstructions": 2197815296,
      "textureSize": [
        64,
        64,
      ],
      "textureSizeValue": 91,
      "textureWrappingFlags": {
        "hFlip": false,
        "hRepeat": true,
        "hStretch": false,
        "vFlip": false,
        "vRepeat": true,
      },
      "textureWrappingValue": 8,
      "vertexColorModeValue": -1,
    },
    {
      "address": 1416,
      "alpha": 1,
      "baseParams": 2147483708,
      "color": [
        1,
        1,
        1,
      ],
      "hasColoredVertices": false,
      "isOpaque": true,
      "polygonDataLength": 136,
      "polygons": [
        {
          "actualVertexCount": 4,
          "address": 1496,
          "flags": {
            "culling": true,
            "cullingType": "front",
            "envMaps": false,
            "gouradShading": false,
            "reuseGlobalParams": false,
            "spriteQuad": true,
            "strip": true,
            "superVertexIndex": false,
            "triangles": false,
          },
          "indices": [
            1,
            0,
            2,
            1,
            2,
            3,
          ],
          "triIndices": [
            1,
            0,
            2,
            1,
            1,
            2,
            3,
            1,
          ],
          "vertexCount": 4,
          "vertexGroupMode": "regular",
          "vertexGroupModeValue": 22,
          "vertices": [
            {
              "address": 1504,
              "addressingMode": "direct",
              "contentAddress": 1504,
              "contentModeValue": 3304718337,
              "index": 0,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                -2000.0001220703125,
                1400,
                499.99969482421875,
              ],
              "uv": [
                0,
                -1.401298464324817e-45,
              ],
            },
            {
              "address": 1536,
              "addressingMode": "direct",
              "contentAddress": 1536,
              "contentModeValue": 3304718337,
              "index": 1,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                -2000.0001220703125,
                1400,
                -2000,
              ],
              "uv": [
                10,
                -1.401298464324817e-45,
              ],
            },
            {
              "address": 1568,
              "addressingMode": "direct",
              "contentAddress": 1568,
              "contentModeValue": 1157234689,
              "index": 2,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                2000.0001220703125,
                1400,
                499.99969482421875,
              ],
              "uv": [
                0,
                -10.000000953674316,
              ],
            },
            {
              "address": 1600,
              "addressingMode": "direct",
              "contentAddress": 1600,
              "contentModeValue": 1157234689,
              "index": 3,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                2000.0001220703125,
                1400,
                -2000,
              ],
              "uv": [
                10,
                -10.000000953674316,
              ],
            },
          ],
        },
      ],
      "position": [
        0,
        0,
        -1000.0003051757812,
      ],
      "specularAlpha": 0,
      "specularColor": [
        1,
        1,
        1,
      ],
      "textureColorFormat": "RGB555",
      "textureColorFormatValue": 8,
      "textureControlValue": 134217728,
      "textureIndex": 0,
      "textureInstructions": 2197815296,
      "textureSize": [
        64,
        64,
      ],
      "textureSizeValue": 91,
      "textureWrappingFlags": {
        "hFlip": false,
        "hRepeat": true,
        "hStretch": false,
        "vFlip": false,
        "vRepeat": true,
      },
      "textureWrappingValue": 8,
      "vertexColorModeValue": -1,
    },
    {
      "address": 1632,
      "alpha": 1,
      "baseParams": 2147483708,
      "color": [
        1,
        1,
        1,
      ],
      "hasColoredVertices": false,
      "isOpaque": true,
      "polygonDataLength": 136,
      "polygons": [
        {
          "actualVertexCount": 4,
          "address": 1712,
          "flags": {
            "culling": true,
            "cullingType": "front",
            "envMaps": false,
            "gouradShading": false,
            "reuseGlobalParams": false,
            "spriteQuad": true,
            "strip": true,
            "superVertexIndex": false,
            "triangles": false,
          },
          "indices": [
            1,
            0,
            2,
            1,
            2,
            3,
          ],
          "triIndices": [
            1,
            0,
            2,
            1,
            1,
            2,
            3,
            1,
          ],
          "vertexCount": 4,
          "vertexGroupMode": "regular",
          "vertexGroupModeValue": 22,
          "vertices": [
            {
              "address": 1720,
              "addressingMode": "direct",
              "contentAddress": 1720,
              "contentModeValue": 3304718337,
              "index": 0,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                -2000.0001220703125,
                0,
                -2000,
              ],
              "uv": [
                0,
                -1.401298464324817e-45,
              ],
            },
            {
              "address": 1752,
              "addressingMode": "direct",
              "contentAddress": 1752,
              "contentModeValue": 3304718337,
              "index": 1,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                -2000.0001220703125,
                1400,
                -2000,
              ],
              "uv": [
                10,
                -1.401298464324817e-45,
              ],
            },
            {
              "address": 1784,
              "addressingMode": "direct",
              "contentAddress": 1784,
              "contentModeValue": 1157234689,
              "index": 2,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                2000.0001220703125,
                0,
                -2000,
              ],
              "uv": [
                0,
                -10.000000953674316,
              ],
            },
            {
              "address": 1816,
              "addressingMode": "direct",
              "contentAddress": 1816,
              "contentModeValue": 1157234689,
              "index": 3,
              "normals": [
                -1,
                0,
                0,
              ],
              "position": [
                2000.0001220703125,
                1400,
                -2000,
              ],
              "uv": [
                10,
                -10.000000953674316,
              ],
            },
          ],
        },
      ],
      "position": [
        0,
        0,
        -1000.0003051757812,
      ],
      "specularAlpha": 0,
      "specularColor": [
        1,
        1,
        1,
      ],
      "textureColorFormat": "RGB555",
      "textureColorFormatValue": 8,
      "textureControlValue": 134217728,
      "textureIndex": 0,
      "textureInstructions": 2197815296,
      "textureSize": [
        64,
        64,
      ],
      "textureSizeValue": 91,
      "textureWrappingFlags": {
        "hFlip": false,
        "hRepeat": true,
        "hStretch": false,
        "vFlip": false,
        "vRepeat": true,
      },
      "textureWrappingValue": 8,
      "vertexColorModeValue": -1,
    },
  ],
  "position": [
    0,
    590.4868774414062,
    -1039.7955322265625,
  ],
  "radius": 2994.93701171875,
  "ramAddress": 0,
  "totalVertexCount": 28,
}
`;
